Add k_usleep() for microsecond thread sleeps.

gnuless opened the pull request:
This is a fixed up and re-submitted version of #16075.

The Kconfig option to enable k_usleep() has been removed.  instead, the documentation for the function has been augmented with caveats for the end user, which serves the same purpose without cluttering the config space.

Commit:
kernel/sched.c: refactor k_sleep() implementation for varied timescales

Current z_impl_k_sleep() does double duty, converting between units
specified by the API and ticks, as well as implementing the sleeping
mechanism itself. This patch separates the API from the mechanism,
so that sleeps need not be tied to millisecond timescales.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
kernel/sched.c: add k_usleep() API function

Add k_usleep() API, analogous to k_sleep(), excepting that the argument
is in microseconds rather than milliseconds.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

jukkar commented on tests/kernel/usleep/src/main.c:
```diff
@@ -0,0 +1,66 @@
+		++retries;
+		start_ms = k_uptime_get();
+
+		for (i = 0; i < LOOPS; ++i)
```
braces need to be added to even one liners

gnuless replied:
@jukkar I'm happy to add braces here, but can you point out the style guideline that says they're required?  The Linux kernel guidelines specifically request that braces NOT be here, according to @ceolin they're not required by MISRA, and the Zephyr Contribution Guidelines say braces are required for `if` and `else` clauses, but says nothing about `for` that I can see.

gnuless replied:
Hmph, I must have misinterpreted @ceolin because MISRA C:2004, 14.8 *does* require the braces.  I'll put them in.

carlescufi replied:
https://docs.zephyrproject.org/latest/contribute/index.html#coding-style

>Add braces to every if and else body, even for single-line code blocks. Use the --ignore BRACES flag to make checkpatch stop complaining.

Doesn't mention `for` loops but it's implied.

gnuless commented:
@andyross I wonder if you might be able to help me get this test through the QEMU CI ...

andrewboie commented on tests/kernel/usleep/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+tests:
+  kernel.usleep:
+    tags: sleep
+    arch_whitelist: x86
```
seems a restrictive whitelist/platform_exclude list.
Is there something which prevents this from working on other arches?

gnuless replied:
The platform_exclude list comes from the x86 boards that don't have an HPET.  The local APIC timer driver isn't up-to-date and doesn't work properly in TICKLESS mode, which is why they're excluded.  (I'll be addressing the local APIC timer driver problems, but that's a separate matter.)

As for the arch restriction - there's some discussion in #16075 about why these sleeps are a bit touch-and-go on various platforms.  For example, the test, as written, will not work on those platforms (the nRFs?) with a 32.768kHz timer source, since test selects a tick of 10us, which is below the hardware timer's resolution.  (I got that number -- 10us -- from a requirement in a project in my lap, so it's not entirely arbitrary.)

In order for the test to be meaningful, we should be testing platforms near the limits of their hardware timers' resolutions, but of course that means tuning the test for various platforms in the CI.  The [admittedly] lazy part of me thinks that's entirely too much work when all *I* need is for this to work right now on x86 with an HPET, and after all, I wrote in the docs that this API is dodgy, right?  On the other hand, as you are rightly pointing out, that's probably being a bit *too* lazy.  I could use some help here figuring out the right balance between the need for good test coverage and the investment of time required...

gnuless replied:
I've rewritten the test to be agnostic about absolute values, so I've removed the restrictions on the test.

tyrantlizard77 reviewed (approved):
code looks fine, curious about the test case filtering

carlescufi commented:
@gnuless thanks for the PR. Can you please take a look at the failing CI?

andyross commented on tests/kernel/usleep/src/main.c:
```diff
@@ -0,0 +1,79 @@
+			 * the conversion from 1us to ticks is rounded
+			 * up to the nearest tick boundary, and sleeps
+			 * always have _TICK_ALIGN (currently 1) added
+			 * to their durations.
```
Hm.... that shouldn't be correct.  The _TICK_ALIGN nonsense is there to make sure we always suspend for at least the requested delay.  In effect it's there to correct a off-by-one error in the way we originally specified the clock driver timeout argument (and which I didn't try to fix during the rework because the timing internals were originally run against the older drivers before the drivers got rewritten).  But in the case where our delays are in microseconds, it shouldn't be doubling the (much larger) tick count.  The intent of the API is that you get the "next tick boundary after the requested delay".  If that's not what's actually happening, this sounds like a bug.

gnuless replied:
Well, the comment is perhaps a bit misleading.  It's a worst of worst cases.

`_TICK_ALIGN` is added because we have no idea how soon in the future the next tick will arrive, so we assume the next tick is imminent, i.e, a sleep of 1 tick will sleep for almost 0 time.  This technique adds in an uncertainty of 0-1 ticks to the intended sleep duration, on average, 0.5 ticks.  The worst case of 1 tick occurs when we sleep in synchronization with the system clock, and that's precisely what happens in this test case: we're going to sleep immediately after the last tick woke us up (guaranteed on every iteration but the first).

Since we have to assume that `_TICK_ALIGN` adds a delay of 0, then we have to round up the provided sleep time to at least a tick boundary, lest we return from the sleep earl.  Hence the use of `ceiling_fraction()` in the division.  The smaller the requested sleep time is relative to the tick duration, the more pronounced the error introduced by this rounding.  In the test case, `k_usleep()` is purposely executing a sleep at the limits of the hardware resolution, showing the most pronounced distortion of the duration.

So these two factors in combination make this test case a really poor illustration of the actual behavior of `k_usleep()` et al.  When we're not synchronized to the system clock and using sleep times that are not near the limits of the hardware clock, the behavior is much more reasonable. For example, with a tick duration of 10us and a requested sleep of 100us, we should see an average sleep time of 105us.  Not so bad.

There are some other solutions.  For example, we could inspect the timer hardware instead of using a worst-case `_TICK_ALIGN`: this is trickier than it might seem at first because of possible races.  Plus there may be some platforms on which such inspection isn't possible.  Alternatively, we could change the API the allow sleeps to return early.  All in all, though, I think the easiest fix is to simply inform users that they should use  a `CONFIG_SYS_CLOCK_TICKS_PER_SEC` that is >> their desired uncertainty.  This basically comes at no cost in a `TICKLESS` kernel.

andyross reviewed (commented):
One worry about the interpretation of tick alignment.  I'm hoping that the too-high error bound mentioned earlier and the seemingly-too-low argument k_usleep() are in fact correlated.  If not, this may be tickling a real bug elsewhere.

nashif commented on tests/kernel/usleep/src/main.c:
```diff
@@ -0,0 +1,79 @@
+/*
```
can this test be added to existing sleep test? Adding a test per API call does not scale.

gnuless replied:
Ah, indeed it can.  It was originally separated out because `prj.conf` had a manual setting for `CONFIG_SYS_CLOCK_TICKS_PER_SEC`, but that's not necessary now.  Moved to the existing sleep test.

Commit:
tests/kernel/sleep: add tests for k_usleep() API

Test that k_usleep() allows sleep durations near the limit of what
the platform's tick rate will allow.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>
